File: #123_Best_Time_to_Buy_and_Sell_Stock_III/Solution.java
public class Solution {
    public static int maxProfit(int[] prices) {
        if (prices.length == 0) return 0;
        int[] left = new int[prices.length];
        int[] right = new int[prices.length];
        /* fill in left */
        left[0] = 0;
        int min = prices[0];
        for (int i = 1; i < prices.length; i++) {
        	left[i] = Math.max(left[i - 1], prices[i] - min);
        	if (prices[i] < min) min = prices[i];
        }
        /* fill in right */
        int max = prices[prices.length - 1];
        right[prices.length - 1] = 0;
        for (int i = prices.length - 2; i >= 0; i--) {
        	right[i] = Math.max(right[i + 1], max - prices[i]);
        	if (prices[i] > max) max = prices[i];
        }
        /* combine both */
        int result = 0;
        for (int i = 0; i < prices.length - 1; i++) {
        	if (i == 0) result = right[i];
        	else result = Math.max(result, left[i] + right[i + 1]);
        }
        result = Math.max(result, left[0]);
        return result;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[] test = {1, 4, 5, 7, 6, 3, 2, 9};
    	System.out.println(sol.maxProfit(test));
    }
}

File: #123_Best_Time_to_Buy_and_Sell_Stock_III/Solution.py
class Solution:
    # @param {integer[]} prices
    # @return {integer}
    def maxProfit(self, prices):
    	if len(prices) < 2:
    		return 0
    	dp0 = [0] * len(prices)
    	minVal = prices[0]
    	for x in range(1, len(prices)):
    		dp0[x] = max(dp0[x - 1], prices[x] - minVal)
    		minVal = min(minVal, prices[x])
    	dp1 = [0] * len(prices)
    	maxVal = prices[len(prices) - 1]
    	for x in range(len(prices) - 2, -1, -1):
    		dp1[x] = max(dp1[x + 1], maxVal - prices[x])
    		maxVal = max(maxVal, prices[x])
    	re = 0;
    	for x in range(len(prices) - 1):
    		re = max(re, dp0[x] + dp1[x + 1])
    	return max(re, dp0[len(prices) - 1], dp1[0])


sol = Solution()
t = [3, 5, 7, 2, 5]
print sol.maxProfit(t)